fl4health/model_bases/pca.py

```diff
@@ -0,0 +1,166 @@
+            else:
+                log(INFO, "Performing reduced SVD on data matrix.")
+            _, S, Vh = torch.linalg.svd(X_prime, full_matrices=self.full_svd)
+            principal_components = Vh.T
+        singular_values = S
+        return principal_components, singular_values
+
+    def maybe_reshape(self, X: Tensor) -> Tensor:
+        """
+        Reshape input tensor X as needed so SVD can be computed.
+        """
+        if len(X.size()) == 2:
+            return torch.squeeze(X.float())
+        else:
+            dim0 = X.size(0)
+            return torch.squeeze(X.view(dim0, -1).float())
+
+    def set_data_mean(self, X: Tensor) -> None:
+        """
+        The primary purpose of this method is to store the mean of the
+        training data so it can be used to center validation/test data later,
+        if needed.
+        """
+        self.data_mean = torch.mean(X, dim=0)
+
+    def centre_data(self, X: Tensor) -> Tensor:
+        assert self.data_mean is not None
+        return X - self.data_mean
+
+    def prepare_data_forward(self, X: Tensor, center_data: bool) -> Tensor:
+        """
+        Prepare input data X for PCA by reshaping and centering it as needed.
+        """
+        X = self.maybe_reshape(X)
+        if center_data:
+            self.set_data_mean(X)
+            return self.centre_data(X)
+        else:
+            return X
+
+    def project_lower_dim(self, X: Tensor, k: Optional[int] = None) -> Tensor:
+        """
+        Project input data X onto the top k principal components.
+
+        Args:
+            X (Tensor): Input Data.
+            k (Optional[int], optional): The number of principal components
+            onto which projection is done. If none, then all principal components will
+            be used in the projection. Defaults to None.
+
+        Returns:
+            Tensor: Projection result.
+        """
+        X_prime = self.maybe_reshape(X)
+        if k:
+            return torch.matmul(X_prime, self.principal_components[:, :k])
+        else:
+            return torch.matmul(X_prime, self.principal_components)
+
+    def project_back(self, X_lower_dim: Tensor) -> Tensor:
+        """
+        Project low-dimensional principal representations back into the original space
+        to recover the reconstruction of data points.
+        """
+        X_lower_dim_prime = self.maybe_reshape(X_lower_dim)
+        k = X_lower_dim.size(1)
+        return torch.matmul(X_lower_dim_prime, self.principal_components[:, :k].T)
+
+    def compute_reconstruction_error(self, X: Tensor, k: Optional[int]) -> float:
+        """
+        Compute the reconstruction error of X under PCA reconstruction.
+
+        More precisely, if X is n by d, and U is the matrix whose columns are the
+        k principal components of X (thus U is d by k), then the reconstruction loss
+        is defined as
+            1 / n * | X @ U @ U.T - X| ** 2.
```

Comment:
Is the `@` symbol here denoting matrix multiplication? If this is standard, we can leave it, I'm just not used to the notation.

Reply:
Yes, `@` is matrix multiplication. I thought it's standard since you can use it to do matrix multiplication in both NumPy and PyTorch but maybe it's not commonly used...?

Reply:
That's good with me. I didn't realize that their docs used this symbol for matrix multiplication. Good to know!